source/common/quic/envoy_quic_dispatcher.cc

```diff
@@ -177,13 +177,24 @@ void EnvoyQuicDispatcher::closeConnectionsWithFilterChain(
     // Retain the number of connections in the list early because closing the connection will change
     // the size.
     const size_t num_connections = connections.size();
+    bool delete_sessions_immediately = false;
     for (size_t i = 0; i < num_connections; ++i) {
       Network::Connection& connection = connections.front().get();
       // This will remove the connection from the list. And the last removal will remove connections
       // from the map as well.
       connection.close(Network::ConnectionCloseType::NoFlush);
+      if (!delete_sessions_immediately && dynamic_cast<QuicFilterManagerConnectionImpl&>(connection)
+                                              .extend_filter_manager_lifetime()) {
+        delete_sessions_immediately = true;
+      }
     }
     ASSERT(connections_by_filter_chain_.find(filter_chain) == connections_by_filter_chain_.end());
+    if (delete_sessions_immediately) {
+      // Explicitly destroy closed sessions in the current call stack, because upon returning the
```

Comment:
So it sounds like we had a lifetime issue where the filter chain configs were getting destroyed in-stack and the sessions were deleted after.
I'm granting that from the test this works, but do you think it might be safer to instead of deleting the sessions online to delay filter chain configs?  inline deletion can definitely be problemaric in Envoy code, but I assume this is mostly quiche code and I'll trust you if you think it's safe.

Reply:
> So it sounds like we had a lifetime issue where the filter chain configs were getting destroyed in-stack 

Unfortunately filter chain configs are already [defer deleted](https://github.com/envoyproxy/envoy/blob/a699868c39fc1da067e9a442e391f1cd2694f175/source/extensions/listener_managers/listener_manager/connection_handler_impl.cc#L243) using Envoy's DeferredDeletable interface and [post()](https://github.com/envoyproxy/envoy/blob/a699868c39fc1da067e9a442e391f1cd2694f175/source/extensions/listener_managers/listener_manager/listener_manager_impl.cc#L854). I don't know if we can defer them even further. The problem is that DeferredDeletable objects + post() will call the destructors at a different time from QuicAlarm scheduled for the next iteration. The latter is how QUIC deletes connections. And with this PR, filter manager is also deleted when the alarm fires. Both QUIC and Envoy are defer-deleting their objects, but just using different mechanism, and their order is not properly coordinated. I think a possible more robust approach is to use weak_pointer in HCM to its config instead of a reference. But that would require more interface change.